src/dtscalibration/io/apsensing.py

```diff
@@ -411,3 +423,263 @@ def metakey(meta, dict_to_parse, prefix):
+
+
+def parse_tra_numbers(val: str):
+    """parsing helper function used by function read_single_tra_file() to determine correct datatype of read string.
+
+    Parameters
+    ----------
+    val : str
+        String value of tra file item
+
+    Returns:
+    --------
+    val : Value in correct datatype (boolean, int, float and string supported),
+    """
+    if val == "True":
+        return True
+    if val == "False":
+        return False
+    if val.isdigit():
+        return int(val)
+    try:  # a bit ugly, but sadly there is no string.isfloat() method...
+        return float(val)
+    except ValueError:
+        return val
+
+
+def read_single_tra_file(tra_filepath):
+    """
+    Using AP Sensing N4386B both POSC (.xml) export and trace (.tra) export can be used to log measurements.
+    This function reads the .tra data and appends it to the dask array, which was read from the POSC export (.xml) file.
+
+    .tra files contain different data then the .xml files from POSC export
+        - more metadata
+        - log_ratio and loss(attenuation) calculated by device
+        - PT100 sensor data (optional only if sensors are connnected to device)
+
+
+    Parameters
+    ----------
+    tra_filepathlist : list of str
+        List of paths that point the the .tra files
+    Notes:
+    ------
+    more metadata could be read from the .tra file and stored in the dask array
+
+    Returns:
+    --------
+    data_dict : dict containing time series measured fibre data by distance
+                                PT100 reference as float
+                                timestamp data
+                                other metadata
+    """
+
+    with open(tra_filepath) as f:
+        file = f.readlines()
+
+    data = [line for line in file if line != "\n"]  # drops out empty lines
+
+    data_dict = {}
+
+    current_section = None
+    for line_with_break in data:
+        line = line_with_break.replace("\n", "")  # drops out linebreaks
+        if line.startswith("["):  # detects new section and sets it as current section
+            current_section = line.replace("[", "").replace("]", "")
+            data_dict[current_section] = {}
+        else:
+            content = line.split(";")
+            content = [parse_tra_numbers(val) for val in content]
+            content_name = content[0]
+            if (
+                len(content) == 2
+            ):  # = metadata & data after trace data (optional sensors and time stamp)
+                data_dict[current_section][content_name] = content[1]
+            else:  # == trace data containing distance, temperature, logratio, attenuation
+                data_dict[current_section][content_name] = tuple(content[1:])
+
+    trace_key = [key for key in data_dict if "Trace." in key][
+        0
+    ]  # find key of trace n in "Trace.n" is unknown
+    data_dict["trace_key"] = trace_key
+
+    return data_dict
+
+
+def append_to_data_vars_structure(data_vars, data_dict_list):
```

Comment:
This function doesn't feel very optimal. Lots of appending and modifying the numpy arrays... But I don't have time now to look into it deeper.

Reply:
@davechris 

I rewrote this function. Should be faster and more legible now :)

```py
def append_to_data_vars_structure(data_vars, data_dict_list):
    """
    append data from .tra files to data_vars structure.
    (The data_vars structure is later on used to initialize the x-array dataset).


    Parameters
    ----------
    data_vars : dictionary containing *.xml data
    data_dict_list: list of dictionaries
                each dictionary in the list contains the data of one .tra file

    Returns:
    --------
    data_vars : dictionary containing *.xml data and *.tra data

    """
    # First derive numer of measurements in each file (should be constant):
    data_dict = data_dict_list[0]
    tr_key = data_dict["trace_key"]
    n_measurements = len([key for key in data_dict[tr_key] if isinstance(key, int)])
    
    # Initialize arrays
    distances = np.zeros((len(data_dict_list), n_measurements))
    t_by_dts = np.zeros((len(data_dict_list), n_measurements))
    log_ratio = np.zeros((len(data_dict_list), n_measurements))
    loss = np.zeros((len(data_dict_list), n_measurements))

    for idx, data_dict in enumerate(data_dict_list):
        # first get distance, t_by_dts, log_ratio and loss as list from dictionary
        tr_key = data_dict["trace_key"]
        data_keys = [key for key in data_dict[tr_key] if isinstance(key, int)]
        data = np.array([data_dict[tr_key][key] for key in data_keys]).T

        # Fill in pre-initialized array
        distances[idx] = data[0]
        t_by_dts[idx] = data[1]
        log_ratio[idx] = data[2]
        loss[idx] = data[3]

    # add log_ratio and attenaution to data_vars
    data_vars["log_ratio_by_dts"] = (("time", "x"), log_ratio)
    data_vars["loss_by_dts"] = (("time", "x"), loss)

    # add reference temp data, if they exist
    for idx_ref_temp in range(1, 5):
        if f"Ref.Temperature.Sensor.{idx_ref_temp}" in data_dict[tr_key]:
            ref_temps = []
            for _, data_dict in enumerate(data_dict_list):
                tr_key = data_dict["trace_key"]
                ref_temps.append(
                    data_dict[tr_key][f"Ref.Temperature.Sensor.{idx_ref_temp}"]
                )
            data_vars[f"probe{idx_ref_temp}Temperature"] = (("time",), ref_temps)

    # check if files match by comparing timestamps and dts temperature
    for idx_t in range(0, len(data_dict_list)):
        # check timestamps
        data_dict = data_dict_list[idx_t]
        tr_key = data_dict["trace_key"]
        dd_ts = pd.Timestamp(
            int(data_dict[tr_key]["Date.Year"]),
            int(data_dict[tr_key]["Date.Month"]),
            int(data_dict[tr_key]["Date.Day"]),
            int(data_dict[tr_key]["Time.Hour"]),
            int(data_dict[tr_key]["Time.Minute"]),
            int(data_dict[tr_key]["Time.Second"]),
        )

        err_msg = f"fatal error in allocation of .xml and .tra data.\nxml file {data_vars['creationDate'][1][idx_t]}\ntra file {str(dd_ts)}\n\n"
        if not data_vars["creationDate"][1][idx_t] == dd_ts:
            raise Exception(err_msg)

        # check dts temperature
        for idx_x in [0, 2, 5]:
            if not data_vars["tmp"][1][idx_x][idx_t] == t_by_dts[idx_t][idx_x]:
                # fatal error in allocation of .tra and .xml data
                raise Exception(err_msg)
    return data_vars

```

Reply:
Thanks for rewriting the function!

Reply:
I am questioning whether we need the trace data at all. 
- distances are not even transferred to data_vars
- t_by_dts is only used for a check (which is a bit unnecessary as we already check the time)
- log_ratio and loss are data we do not really need. Originally I was interested in comparing them do the self-calibrated data, but I never did it and feel that it is unnecessary, as I see already on the temperature curve, that the device calibration is not that good.

Reply:
I will add a flag `load_full_tra = False` to read_apsensing_files. Thus, unnecessary trace data will not be imported in the default case.